### DISCUSSION TITLE
How to set timezone in time relate function?

### DISCUSSION THREAD

**Discussion by water5** [micropython/micropython#16044] [General]
Can I use timezone in MicroPython, how to set the offset?
Is it use `RTC.init()` `tzinfo` argument? it's not describe in document.

**Comment by peterhinch** [comment]
MicroPython does not support timezones.

It is usual to set the clock on a microcontroller to local time. In a location with DST, a typical long running application will run winter time. If it is required to display local time the application performs the conversion for summer time.

**Comment by GM-Script-Writer-62850** [comment]
i made my own way to do this and dst
https://github.com/GM-Script-Writer-62850/PICO_W_Thermostat/blob/main/PICO/lib/TIME.py (i missed time.ticks_ms BTW)

https://github.com/GM-Script-Writer-62850/PICO_W_Thermostat/blob/main/PICO/main.py
note line 14 and the getSolar() function on line 222

**Comment by bulletmark** [comment]
It is convention that computers store universal time (i.e. UTC/GMT) in the hardware clock. When the time gets displayed you either display that UTC time, or if you want to display local time (i.e. time offset by your local timezone), then you need a timezone offset which you will have to configure in your system/app. I use the standard MicroPython [datetime](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/datetime/datetime.py) library which includes a [`class timezone(tzinfo)`](https://github.com/micropython/micropython-lib/blob/68e3e07bc7ab63931cead3854b2a114e9a084248/python-stdlib/datetime/datetime.py#L247C1-L247C22) implementation so you can set your local `tzinfo` using that time offset. If instead you want to use any arbitary time zone name, e.g. "Australia/Sydney" which changes time zone twice per year, then you will have to periodically fetch the timezone from somewhere. I use a background service on ESP32 to periodically sync the clock to UTC time using [`ntptime.settime()`](https://docs.micropython.org/en/latest/esp8266/quickref.html#real-time-clock-rtc) and I get the offset for timezone name (at a slower rate) from the free API at http://worldtimeapi.org/.

**Reply by GM-Script-Writer-62850** [reply]
for DST on my server i use this in PHP to send the start/end of DST for the current year then i update that when the year rolls over
```
function getDST(){
        $timeZone = new DateTimeZone('America/New_York');
        $trans = $timeZone->getTransitions(strtotime('first day of January this year'));
        return (object)array(
                "start" => $trans[1]["ts"],
                "end" => $trans[2]["ts"],
                "update" => strtotime('first day of January next year')+7201// 2AM
        );
}
```
here is the output:
```
object(stdClass)#2 (3) {
  ["start"]=>
  int(1710054000)
  ["end"]=>
  int(1730613600)
  ["update"]=>
  int(1735696801)
}
```
start is when DST begins and end is when it is done with, update is when the new values can be retrieved

**Comment by bidrohini** [comment]
If you're using a board like the ESP8266 or ESP32 with internet access, you can use ntptime to get the current UTC time from an NTP server and then adjust it for your timezone.
